WIP: Make mutating immutables easier

Keno opened the pull request:
I like immutables, I really do, they're simple to work with, fast, don't cause allocations. Really the only thing that bothers me about them is that they're well, immutable, making them a bit of pain to work with, esp. when wanting to construct one incrementally. So here's an attempt to remedy that.
Example usage:

```
julia> struct Foo
       a::Int
       end

julia> x = Foo(1)
Foo(1)

julia> x@a = 2
Foo(2)

julia> x
Foo(2)
```

The ways this works is that under the hood, it creates a new immutable object with the specified field modified and then assigns it back to the appropriate place. Syntax wise, everything to the left of the `@` is what's being assigned to, everything to the right of the `@` is what is to be modified. E.g.

```
struct Foo4; d::Int; end
struct Foo3; c::Foo4; end
mutable struct Foo2; b::Foo3; end
struct Foo1; a::Foo2; end
x = Foo1(Foo2(Foo3(Foo4(1))))
x.a.b@c.d = 2
julia> x
Foo1(Foo2(Foo3(Foo4(2))))
```

Internally, everything to the right of the `@` gets lowered to `setindex` (no bang) and `setfield` (also no bang), which are overridable by the user. The intent is to disallow `setfield` for immutables that have a non-default inner constructor, allowing the user to provide their own which checks any required invariants. That part isn't implemented here yet, however.

Lastly, LLVM isn't currently too happy about the IR this generates, so I'm working on making that happen to make sure this actually performs ok. I think from the julia side, this is pretty much the extent of it though. Pretty much untested at the moment. I want to get the LLVM side of things done first. With that in mind, feel free to check out this branch and see if you like it.

One motivating example here is of course efficient, generic fixed size arrays, so here's an example of that:
```
mutable struct MArray{Size, T, D, L} <: AbstractArray{T, D}
    data::NTuple{L, T}
end

Base.size(::MArray{Size}) where {Size} = tuple(Size.parameters...)
@inline to_linear_index(::Type{Tuple{}}, idxs, stride) = 1
@inline to_linear_index(::Type, idxs::Tuple{}, stride) = 1
@inline to_linear_index(::Type{Tuple{A}}, idxs, stride) where {A} = idxs[1]
@inline to_linear_index(::Type{Tuple{A,B}}, idxs, stride) where {A, B} = A*(idxs[1]-1) + idxs[2]
@inline to_linear_index(T::Type{<:Tuple{A, Vararg{Any}}}, idxs, stride) where A =
    (stride * (idxs[1]-1)) + to_linear_index(Base.tuple_type_tail(T), idxs[2:end], stride * A)
function Base.setindex!(a::MArray{Size}, v, idxs...) where Size
    @inbounds a.data@[to_linear_index(Size, idxs, 1)] = v
    v
end
Base.getindex(a::MArray{Size}, idxs...) where {Size} = a.data[to_linear_index(Size, idxs, 1)]

a = MArray{Tuple{2, 2}, Int, 2, 4}((1,2,3,4))
b = [1 2; 3 4]
c = [4 5; 6 7]


# Look ma, no allocations
julia> @benchmark A_mul_B!(a, b, c)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     86.193 ns (0.00% GC)
  median time:      86.993 ns (0.00% GC)
  mean time:        87.978 ns (0.00% GC)
  maximum time:     3.282 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     961
```

Fixes #11902
Supersedes #12113

Keno committed:
Add runtime support for non-mutating setindex

Keno committed:
Add basic parsing and lowering support

Keno committed:
Codegen and inference

JeffBezanson commented:
Wonderful! This approach has my full approval.

JeffBezanson commented on test/parse.jl:
```diff
@@ -1187,3 +1187,14 @@ module Test21607
 end
+
+# Basic parsing for in place assign
+@test parse("a@b = 1") == :(a@b = 1)
```
This kind of test isn't really effective; for example it would pass if `a@b = 1` parsed as `42`.

Keno replied:
Primarily I wanted to make sure it doesn't throw, because I'm not sure the parsing is final yet. I also find it useful, because that way I can use parse.jl to test JuliaParser.

yuyichao commented:
Another approach I was thinking about that is related is a more general representation of memory location, i.e. `Ref`. The motivation is to support better atomics. The C/C++ model of expressing this doesn't work for us very well since we are missing the notion of pointer or reference which are the input type of atomic intrinsics in C/C++. If we want to bind atomic operations to certain types, we'll have to add additional indirection (what we do now) or make them a special case in the system which carries their location information (I don't think this is doable). That's why I think we should define atomic operations on `Ref`s which is essentially our representation of memory location. For example, given `struct B x::Int; y::Int end` `a::Vector{B}` one would do atomic operation on the `a[1].x` element (a `Int`) by `atomic_op!(reffield(refindex(a, 1), :x))`.

It seems that most of the problems here can be solved with a similar lowering so `a@[1].x = v` can be lowerd to `unsafe_store!(reffield(refindex(a, 1), :x), v)`. This won't cover the case of mutating a immutable variable by returning a new one though that can probably be distinguished from parser level.

I'm still looking for alternative proposal from atomic ops but haven't found any working ones yet. If we are going to have better atomic ops support (I think we should) it'll be nice if we can unify the API.

Keno commented:
I had originally proposed that as an implementation strategy, but people didn't like it because the lack of access protection for immutables that create constraints in their inner constructors.

yuyichao commented:
I feel like with `reinterpret(::Array, ...)` it shouldn't be that big of a concern........ Just for this particular problem the approach here looks fine but I really can't come up with a better approach for atomics though...

Keno commented:
Well, for what it's worth, my original proposal was to have a
`immutable RefField{Base, Derived}; a::Int; end`
type that would get created appropriately, so that's still a possibility. People just seemed really opposed to being able to bypass constructor invariants.

JeffBezanson commented:
> I feel like with reinterpret(::Array, ...) it shouldn't be that big of a concern

One sketchy feature doesn't make it ok to add more sketchy features. The uses of `reinterpret` are also quite limited, and the function could probably be restricted much more than it is now. This mutation feature, on the other hand, is a much more general feature that is likely to be used heavily.

yuyichao commented:
I was thinking of `struct RefField{Eltype} a; offset::Int end` so that `Int` field from different source can have the same type. (and to reduce the number of types in general) but I guess that detail can be decided later.

yuyichao commented:
> One sketchy feature doesn't make it ok to add more sketchy features.

Sure. I don't feel like eitheer is very sketchy though. There are invariants that the compilier can use and the ones that users can use. The constructor invariant is something only the user can use so it wouldn't create any undefined behavior if broken. Also, if one decides to mutate a field, he is already touching internal representation (unless it is documented to be public) so it isn't anything different from what we do for mutable struct either, i.e. we document that fields are in general private and mutating them can break other code but if it's needed for debugging or testing then you can do it without having the compiler stop you from doing what should be possible to do.

Keno commented:
One possibility is to use this lowering for the LHS of an assign, but use the `RefField` lowering for this syntax every where else. I feel like a binary externally constrained/externally unconstrained would be ok. E.g.:

```
struct MyConstrained
counter::Int
launch_nukes::Bool
MyConstrained() = MyConstrained(counter, false)
function gepfield(x::MyConstrained, field)
field == :x && return Core.gepfield(x, field)
error("nice try")
end
end
```

yuyichao commented:
Having a way to disallow ref (gep) of a certain field would certainly be possible. It will also stop that field from any useful mutation though, which is why I didn't bring that up....

Keno commented:
Well, seems like you could allow writing an inner method like so:
```
struct MyConstrained
    counter::Int
    launch_nukes::Bool
    MyConstrained() = MyConstrained(counter, false)
    function gepfield(x::MyConstrained, field)
        field == :x && return Core.gepfield(x, field)
        error("nice try")
    end
    function launch!(x::RefField{MyConstrained}, code)
        authorized(code) || error("This incident will be reported")
        Core.gepfield(x, :launch_nukes) = true
    end
end
```

yuyichao commented:
For this one, yes. It'll be quite complicated for atomics though. Maybe just pass in the value during `gep` and let the caller do the storing?

yuyichao commented:
Or at least separate the checking from the actual memory op.

Keno commented:
Well, I just figured either you treat a field as a general atomic in which case it's either all or nothing (and can be done via gepfield) or you have to do the memory modification yourself (by defining an inner method).

yuyichao commented:
So a type with checked invariance has to define all memory ops on the field it needs checked atomics? This would also include the case where the atomic is done only on a subfield?

Keno commented:
Well, if they subfield itself has constraints, then yes, I don't really see how else you can do it. If not, you can define a `gepfield` for that field.

yuyichao commented:
So what about this. If a type allows external mutation of a field, it'll be all or nothing. If it want to check, it needs to provide it's own interface to do that. It can do that by defining a function that accepts a correct `Ref` type. So if we have a `struct A x::Int; y::Int end` and it want to allow normal store to `y` with checked value to ensure `y > x`. It'll define

```julia
function mutate_y(ref, y)
    obj = ref.obj
    @assert y > obj.x
    unsafe_store!(ref, y)
end
```

And the user will call it with `mutaate_y(container@[i].y, new_y)` where the non-assigning `container@[i].y` is a place holder syntax for constructing the ref field object.

yuyichao commented:
Or even

```
function mutate_y(ref_to_obj, y)
    obj = ref.obj
    @assert y > obj.x
    unsafe_store!(Core.fieldref_for_fieldref(ref, :y), y)
end
```

and called with `mutate_y(container@[i])`

Keno commented:
Right, I think the second is ok. Your `fieldref_for_fieldref` is what I called `Core.gepfield`, which may be distinct from `Base.gepfield` called by the syntax (and may also only be available from within type scope like `new`, etc.

yuyichao commented:
> is what I called Core.gepfield

It's slightly different since `Core.gepfield` would take a (mutable) object and construct a gep on the object. `Core.gepfield_for_ref` would take a ref to an (immutable) object and construct a gep on the object.

yuyichao commented:
Of course we can make it dispatch but I don't want to block constructing a field ref on a ref object (`RefValue`) at this low level.

Keno commented:
If you look at my launch_nukes example above, I had implicitly assumed that `gepfield` allows this by dispatch. In fact allowing it is the only way to properly implement the syntax.

Keno commented:
But yeah, I do realize that my gepfield implementation example should have probably taken a RefField.

yuyichao commented:
Ok, yeah, that's right.

yuyichao commented:
I wonder if `gepfield` should take a type explicitly like llvm does to avoid confusion.....................

Keno commented:
So only `gepfield(::Type{T}, RefField{T}, fld)` would be allowed?

yuyichao commented:
`gepfield(::Type{T}, ::Union{Ref{T},T}, fld)`

yuyichao commented:
Or I guess maybe dispatch based on mutability is good enough.

Keno commented:
Right, fair enough. That all seems sensible to me. That does raise the question of what to do about this PR. I do kinda like the current lowering however, so there's a question of if we're ok with this stuff behaving asymmetrically on the lhs and rhs.

Keno commented:
And in particular `(a@b)[] = x` and `a@b[] = x` being different things (though I guess in the default case they could behave the same).

yuyichao commented:
The syntax it's replacing (`setindex!`/`getindex` and `setfield!`/`getfield`) already have asymmetric lowering on rhs and lhs already.....

yuyichao commented:
Also, wouldn't the reffield lowering be easier for llvm to optimize? (obviously we need to optimize out the allocation ourself) Since the memory operation would be directly on the slot instead of on the whole object and relying on llvm to shrink the store size.

JeffBezanson commented:
It's worth noting that the approach currently here --- load an immutable value, make a new one different in one field, and store that back --- is appealing partly because it doesn't involve any truly new operations at all. It's something you can do right now. It simply lacks convenient syntax and compiler support. So this can be seen more as a completion of the existing object model rather than some wacky new thing. But if we do it a different way, that would run the risk of being a wacky new thing.

yuyichao commented:
That's exactly why I brought up atomics. The current approach is invalid for atomics. It is not allowed to perform a wider store and rely on optimization to reduce it. I only think we need a better approach because I think it's worse to have two that serves very similar purpose.

JeffBezanson commented:
Here are three operations: (1) load, (2) store, (3) allocate new struct. Unless you're proposing to remove one of those operations, you can't get away from this existing in the language, you're simply asking for it to have bad syntax. I don't buy that atomic operations should be driving something this fundamental.

Maybe say more about how you'd like to use atomics that requires this? It seems to me they're needed rarely enough that the awkward syntax could be pushed to that case, but maybe I'm missing something.

yuyichao commented:
I'm fine with having atomics using a bad syntax. What's needed for it is an API to operate on custom types (`AbstractArray` subtypes for example). It just happens that such an API can also be used to do what this PR wants to do. I'm not bound to using a good syntax for this. The rvalue `@` syntax above is just to make it more readable in discussion. We don't have to have that.

But that does remind me that this approach is actually fundamentally flawed in multithreaded environment. I do agree that atomics are rare (i.e. synchronizations should be rare) but non-conflicting concurrent data access won't be. It should be allowed to write to different elements in a `StaticArray` from different threads without data race. However, such thing will not be allowed with this approach since the semantics of this approach disallows it.

In order to support multiple thread, an operation that is meant to access only one element is not allowed to touch any other memory **semantically** otherwise it'll create data race. Since this is a semantic issue, it also has to be reflected in the API and cannot be relying on optimization to optimize it out since optimization cannot do it in all cases and especially less reliable in the cases where a user overload any part of the call chain.

JeffBezanson commented:
> It should be allowed to write to different elements in a StaticArray from different threads without data race.

Ah, that's a very interesting point. I think my reaction is that this should *not* be allowed. That's a good example of enforcing the immutability abstraction even without constructor-based invariants! With immutable data, the threads aren't writing to the array at all, they're just computing different arrays, and you shouldn't end up with an array that nobody actually computed.

That example clarifies that we are indeed talking about two different operations, so they probably should be totally separate features.

yuyichao commented:
> That example clarifies that we are indeed talking about two different operations, so they probably should be totally separate features.

Given that the last example in the original post is essentially StaticArray I suppose this is also trying to address that use case?

So assuming we want to have the thread safe version (i.e. partial modification of object), which I hope is the case, is there a use case of overwriting the whole object that can't be covered with this? We already have all three (load,store,allocate) operation in the language with concise syntax already so I think the only thing we are missing is a syntax to combine the three, i.e. load, modify a field, store (incidentally a very common patten in atomics..............) For the user, shouldn't both feature be able to support this usage? For library writing I'm not sure which one is easier but I don't really think the ref interface would be harder to work with.

Keno commented:
There is the argument that concurrent modification of an object from different threads shouldn't be allowed at all without some semantic guarantee that what they're doing is deterministic.

yuyichao commented:
> concurrent modification of an object from different threads shouldn't be allowed at all without some semantic guarantee that what they're doing is deterministic.

Right, and that guarantee comes from the definition of memory location. We currently guarentee that mutation different field of the same mutable object is not a data race, which would agree with the definition in C/C++. I agree it is possible to define it differently but I really don't see the point of making multithreaded programing even harder than it already is and encouraging people to program in a racy way.

And even if we define this in a racy way, it'll still be a regression for StaticArrays if it ever want to use this. People are already hit by this behavior for BitArray (which is semantically racy) https://discourse.julialang.org/t/odd-behavior-from-threads-with-bitwise-and/3191 and this would not be a user error if it happens on an array type that doesn't use bits as storage type.

yuyichao commented:
And needless to say, defining field mutation to be racy on the same object would certainly be very unexpected for anyone from other languages. Defining a syntax that looks and named really similar to field mutation to have the same racy behavior would probably be equally unexpected.

JeffBezanson commented:
> defining field mutation to be racy on the same object

But immutable objects don't support field mutation. If you want multiple threads to write to different elements of an array, you need a mutable array.

yuyichao commented:
For a start, StaticArray is mutable so it still holds true that this is useless for StaticArray.

Immutable objects don't support field mutation but it is also not a single memory location. The syntax is also important here. If the syntax/functionn name here doesn't look like setfield, I'd have no problem implementing it in a racy way. In that sense, yes I am suggesting now that a syntax surgar that is semantically racy should have a worse syntax than something else that can accomplish exactly the same but is not racy and easier to optimize.

yuyichao commented:
Also, the quoted part is replying to "concurrent modification of an object" where the object is mutable and I was just saying we shouldn't change the current memory model.

"Defining a syntax that looks and named really similar to field mutation to have the same racy behavior would probably be equally unexpected." is what applies to immutable object.

JeffBezanson commented:
I find it very biased to call this the "racy" version. Sure, if one thread writes `(1, 0)` somewhere while another thread tries to write `(0, 1)` there, that's a data race, but that's nothing new. Trying to prevent data races in user code is not the topic of this issue. If you have a mutable object with a single tuple field, I would argue that this is the behavior you're asking for. To get multiple mutable locations, we might need to add a repeated-fields feature.

yuyichao commented:
> I find it very biased to call this the "racy" version. Sure, if one thread writes (1, 0) somewhere while another thread tries to write (0, 1) there, that's a data race, but that's nothing new. Trying to prevent data races in user code is not the topic of this issue.

Calling data racy nothing new and encourage it by defining convenient syntax for it is also very biased. FWIW, with the current syntax, if the user is doing `a@[][1] = 1` on one thread and `a@[][2] = 1` on another, he is **not** explicitly writing `(0, 1)` on one thread and `(1, 0)` on another, it's the compiler that does it. And the issue is precisely that the compiler must not implicitly create those additional stores for the user. Multithreaded programming is hard partially because compilers were not designed with the possibility of data race in mind. C/C++ have already moved in the right direction and I don't think we should try our best to make it harder.

> If you have a mutable object with a single tuple field, I would argue that this is the behavior you're asking for. To get multiple mutable locations, we might need to add a repeated-fields feature.

"mutable location" is not the right concept, memory location is. And memory location is essentially something you can access as a single object, which is more precisely defined as any primitive type object.

JeffBezanson commented:
> "mutable location" is not the right concept, memory location is

Memory location is not the right concept in that it violates abstraction.

`a[i] = x` is already convenient syntax for a data race, so yes this is nothing new. But I can understand why you don't like the syntax and behavior proposed here. I'm open to coming up with a different syntax that more closely reflects what is happening.

yuyichao commented:
> Memory location is not the right concept in that it violates abstraction.

It doesn't and that's exactly what https://github.com/JuliaLang/julia/pull/21912#issuecomment-301910308 is about.

> `a[i] = x` is already convenient syntax for a data race

No it's not. This is a store to memory location(s) `a[i]` and no where else. Meaning it's not a race when any of `a` or `i` are different on two threads. That is not the case for the proposed implementation, which is the problem.

> I'm open to coming up with a different syntax that more closely reflects what is happening.

Well, then basically by not having a syntax for this at all, and requires calling `x = setfield(x, i, v)`. It just need to make sure `x =` is explicit so it can also be `x = x@i <- v` I'm not very good at coming up with the RHS of this.

JeffBezanson commented:
One thing I like about `a@x` as an rvalue creating some kind of reference is that it's consistent with `a@[...]` returning an array view.

Keno commented:
Having slept on this, I think I would like to go ahead with the approach that's currently started in this PR for the following reasons:
1. It's easy to explain
2. It's semantically consistent for both mutable and immutable types (if something on the right happens to be a mutable type, it gets copied).
3. It does what we want
4. It doesn't change the current memory model
Primarily I'm uncomfortable with multi-threading driving major semantic decisions at this point. We don't have enough experience with what multi-threading in julia can or should look like. This change is the result of many years of experience with the current system. We don't have similar levels of experience with multi threading. As more people start using multi-threading I think we'll start seeing where the pain points are and at that point, I do think we can make an informed decision here. Making design decisions without having had a chance to explore the design space is probably a bad idea.

yuyichao commented:
> It's easy to explain

I disagree but this is subjective.

> It's semantically consistent for both mutable and immutable types (if something on the right happens to be a mutable type, it gets copied).

Meaning it could silently do something you didn't expect. I actually think it's better to throw an error for mutable.

> It does what we want

Poorly and inefficiently

> It doesn't change the current memory model

It does a lot. Immutable is not a black box type. It has multiple field that is visible to the user. Those are not pieces that cannot be separated together.

> Primarily I'm uncomfortable with multi-threading driving major semantic decisions at this point. We don't have enough experience with what multi-threading in julia can or should look like.

Exactly, that's why we should borrow experiences from C/C++.

> This change is the result of many years of experience with the current system. We don't have similar levels of experience with multi threading. As more people start using multi-threading I think we'll start seeing where the pain points are and at that point, I do think we can make an informed decision here. Making design decisions without having had a chance to explore the design space is probably a bad idea.

The discourse post I linked about already shows that people are actually surprised by data races caused by modifying different part of an array without realising that they belongs to the same memory location. The semantics in C/C++ is also the result of years of experience with multi-threading and how you can painfully correct the mistake made in compiler design due to now having considered this early on. I think it is a bad idea if we do not consider it.

andyferris commented:
I'm extremely excited that this is being looked at :) Since this is a complex issue, I'm going to break up me thoughts a bit (sorry, I know I'm a bit wordy...)

 * I agree with Jeff that it's wonderful to have some concise syntax that preserves all the *current* behaviors surrounding immutability. It's more-or-less just syntax sugar (e.g. *StaticArrays* already defines `setindex` and it seems well optimized by the compiler).
 * I also think @yuyichao concerns are very real and should not be discounted. As an example, say I have `image::Matrix{RGB{UInt8}}` (ignoring fixed-point numbers for the moment), and I want one thread to populate the red channel, another the green, and a third thread for the blue. If I used this syntax I *really* wouldn't be expecting a data race where the red thread writes overwrites the blue channel - I would assume that the language supports writing to the fields directly, more-or-less as the syntax looks. Note I can currently get thread-safe behavior via `reinterpret(UInt8, image, ...)` or even by `ccall` but it's less convenient than `image[i.j]@r = 255`.
 * I feel that having *partially*-enforced invariants for immutables is quite unfortunate - and would suggest either (A) fully locking this down (trust no-one), or else (B) relax and stop worrying about the user getting around it (trust everyone). There may be more than one approach, but to fully lock it down, I've thought of the following: semantically the invariant must be known to be enforced whenever the object has (or could have) a direct, covariant binding / SSAValue / be passed to a function / whatever, e.g. either upon construction (`x = RGB(1,2,3)`) or once the data is unpacked from any `Ref`/`Array`/`Buffer` and brought to a covariant position (more-or-less brought to the stack). *But*, whenever a memory reference to the object, it is free game. On a practical/performance level, inference and codegen might have to do some work to track that the checking of the invariant doesn't happen more often than necessary. 
 * Given the above (A or B), I feel it would be consistent to say that all the memory (and any subpart thereof) referenced to by a `Ref`, etc, is free game to be mutated - semantically in a partial sense or a entire "immutable" object at once.

I think this would give us a few advantages. For example, in that the FFI is pretty central to Julia, we can pass and receive references to objects to C/C++ code which can make arbitrary changes to any piece of referenced memory, yet it would be nice to have their invariants be enforced by Julia, and to have Julia code which can be made as performant and thread-safe as C/C++ code.

My opinion is therefore that this PR might be a good, short-term syntax improvement. But - I honestly feel that if we figure out what we really want to allow in references to immutable data, have a good way of making this performant (e.g. references that can be on the stack, be deterministically GC'd, etc), and have a good story around invariant before making a change like this, then the syntax we want to match that new semantic will become crystal clear.

Keno commented:
> Meaning it could silently do something you didn't expect. I actually think it's better to throw an error for mutable.

Depends on what your expectation is. The semantics are that everything to the right of the add gets copied.

> Poorly and inefficiently

I'm confident I can get LLVM to optimize the redundant copies away.

> Exactly, that's why we should borrow experiences from C/C++.

Julia is not either of those languages though. We don't expect the majority of julia users to write code at the level of C/C++.

> The discourse post I linked about already shows that people are actually surprised by data races caused by modifying different part of an array without realising that they belongs to the same memory location. The semantics in C/C++ is also the result of years of experience with multi-threading and how you can painfully correct the mistake made in compiler design due to now having considered this early on. I think it is a bad idea if we do not consider it.

Yes, which is why I think there's a very real possibility that we want to disallow this kind of concurrent modification from multiple threads and if we do want to allow it have a very constrained interface of doing so.

yuyichao commented:
> Yes, which is why I think there's a very real possibility that we want to disallow this kind of concurrent modification from multiple threads and if we do want to allow it have a very constrained interface of doing so.

??? The post clearly shows that users expect modifing unrelated element concurrently to be allowed and they would be right if it wasn't bitfields. Therefore, we should either allow it or find a way to catch it when it happens.

Keno commented:
> ??? The post clearly shows that users expect modifing unrelated element concurrently to be allowed and they would be right if it wasn't bitfields. Therefore, we should either allow it or find a way to catch it when it happens.

There's a difference between what we do right now and what we want to do eventually. Right now everything is a bit of a memory-modification free for all. I would be much-much happier if most mutable objects were thread-local, with specific well-defined interfaces for memory locations mutable from multiple threads at once.

yuyichao commented:
> There's a difference between what we do right now and what we want to do eventually. Right now everything is a bit of a memory-modification free for all. I would be much-much happier if most mutable objects were thread-local, with specific well-defined interfaces for memory locations mutable from multiple threads at once.

If you want to implement that, sure. But precisely because of that we want something that works like what people would expected (in terms of race condition) before you have that. I don't see why that's a good reason to just give people a harder time.

Keno commented:
> If you want to implement that, sure. But precisely because of that we want something that works like what people would expected (in terms of race condition) before you have that. I don't see why that's a good reason to just give people a harder time.

We can't really make any atomicity guarantees with the assignment syntax anyway though (esp, across platforms). Your argument is that you want to align the syntax with a general way for memory location referencing to be used in our multi threading model, and I'm sympathetic to that. However, what I'm saying is that we may not WANT to be able to express arbitrary threading-safe memory locations in our threading model.

Once we do have a general multi threading model, it'll be a huge change across the ecosystem anyway, so we can revisit whether changing the lowering of this syntax makes sense.

yuyichao commented:
> We can't really make any atomicity guarantees with the assignment syntax anyway though (esp, across platforms).

Why is that. The definition of memory location is **not** platform dependent.

> Once we do have a general multi threading model, it'll be a huge change across the ecosystem anyway, so we can revisit whether changing the lowering of this syntax makes sense.

The point is that it shouldn't be. The nice thing about what C/C++ (and also Java and probably everyone else) uses as their threading model is that you can just write normal code without worrying about multithreading at all. This precisely require the compiler / syntax doesn't add spurious data race for the user automatically / implicitly.

yuyichao commented:
And again the main point is that the syntax should not hide this behavior. As I mentioned previously, as long as the syntax is `x = ...` instead of `x@i = ...` or `x.i = ...` I'll have no problem with it.

Keno commented:
> Why is that. The definition of memory location is not platform dependent.

Well, first because we don't really have a notion of primitive type. We could say all bits types smaller than pointer size (which is also pretty arbitrary, but mostly fine because that's essentially what C++ does). Also, the lowering is pretty arbitrary, so it's not really guaranteed that something that looks like a memory location on the LHS is actually one. Most importantly however, I'm uncomfortable with atomicity not being distinguished at the syntactic level. If the syntax is sometimes atomic and sometimes not, depending on what the LHS happens to be, that seems deeply problematic to me, especially in generic code.

> The point is that it shouldn't be. The nice thing about what C/C++ (and also Java and probably everyone else) uses as their threading model is that you can just write normal code without worrying about multithreading at all. This precisely require the compiler / syntax doesn't add spurious data race for the user automatically / implicitly.

Yeah, but we have an option to be better here. I'm all for having nice ways to do whatever threading support we can, I just don't think that this particular syntax is the correct way to express modification in a threaded context.

yuyichao commented:
> I'm uncomfortable with atomicity not being distinguished at the syntactic level.
> If the syntax is sometimes atomic and sometimes not, depending on what the LHS happens to be, that seems deeply problematic to me, especially in generic code.

So

1. Get rid of/don't use BitArray, it's the only exception we currently have
2. Don't do the lowering here, it's the new exception.

Keno commented:
No, what I'm saying is "don't concurrently modify objects from different threads"

Keno commented:
Any assignment larger than pointer width is non-atomic. My problem is that trying to ascribe semantic atomicity to some operations, when currently we don't attempt to make any such guarantees.

yuyichao commented:
> Any assignment larger than pointer width is non-atomic.

Correct but irrelevant. Any assignment larger than bytes is race-free.

> My problem is that trying to ascribe semantic atomicity to some operations, when currently we don't attempt to make any such guarantees.

I'm not talking about atomicity. (I did bring this up in the context of atomics but haven't be talking about it since https://github.com/JuliaLang/julia/pull/21912#issuecomment-301931968) We shouldn't make any guarantee on the atomicity for any memory ops and that's exactly why I think we need to have atomic version of the operation here (and I'll even be fine to have to use `atomic_load_relaxed(a@[i])` to load an array of small primitive types atomically). The guarantee we do have is race condition, that's the one we shouldn't implicitly break.

Keno commented:
I wonder if what we need is to different types of mutable locations, "memory" and "logical", where for the former, the types are mostly advisable, support reinterpret, gepindex, etc whereas the latter don't. Then, any time you pull an immutable out of a "memory" style location, it has its invariants verified.

Keno commented:
So @JeffBezanson and I had a pretty long discussion about this to try to explore as much of the design space as we can, and we think we may have found a middle ground that balances all the concerns. First, we're splitting this into two features. What this PR does is now this:
```
a = a@(x=1)
# Some more examples here:
a.b = a.b@(c.d=1)
a.b @=(c.d=1)
a = a@([1] = 1)
```
so, e.g. you could write
```
update_green(a::RGB, ng) = a@(g = ng)
```
Additionally, we're introducing what I originally suggested as an implementation strategy and what @yuyichao favors:
```
struct Foo; a::Int; end
x = Ref{Foo{Int}}()
x[]@a = 2
```
That gets lowered to `setindex!(x[]@a, 2)`, where `x[]@a` is syntax for a field reference as discussed above. As a result, this again expands to
```
#1 = gepindex(x)
#2 = gepfield(#1, :a)
setindex!(#2, 2)
```
Where that setindex! essentially does a safe store. Note that you're not allowed to create a `RefField` with an arbitrary index, but only by calling the (default) gepfield function.

Now, back to the new setfield syntax. Rather than the setfield fallback being a builtin, it's now this
```
function setfield{T}(x::T, sym, v)
    mem = Ref{T}(x), sym)
    gepindex(x,sym)[] = v
    mem
end
```
We figured there's sort of a natural progression of implementation difficulty here:
1. No inner constructor -> Free for all for arbitrary modification
2. Inner Constructor to enforce invariants -> Fine, no modifications allowed
3. Same, but want to allow some fields to be modified (e.g. because the constraints are only over some of the fields) -> Fine, need to implement a custom inner gepfield though
4. Same as 2, but want to maintain some complex invariants over part of the object, fine, but you have to implement setfield.

Thoughts?

iamed2 commented:
The semantics of the hidden overwrite in the initial proposal made me uncomfortable but I like the new proposal. Why new syntax and not a function call though?

Keno commented:
The indexing chains have pretty complicated lowering, so they're quite a pain to write manually.

yuyichao commented:
> `a.b @=(c.d=1)`

what's this?

So just to check if I'm un-derstanding this proposal correctly.

1. The `@` syntax without `=` will create a ref and assigning to it is semantically defined as a store to only that field's memory location.
2. The `@` syntax with `=` will create a whole new object (recursively mutating fields). The default implementatiion will be shared with the ref syntax using the default setfield.
3. `x = x@([].a = 2)` could be used instead of `x@[].a = 2` when one of the type disallows gep explicitly
4. I hope it is still possible to emit `x@[].a = 2` for a type that disallow gep for private implentation? (Possibly by calling the Core.gepfield directly)

This sounds good to me. 1 is essentially what I was asking for and the sharing of implementation in this way makes sense to me. I wonder whether the disabling default gep would be implicit on inner constructor since there's also a lot of types that doesn't use inner constructor to enforce invariance. Also what about mutable types. Their constructor cannot provide language level invariance. Would constructing ref on them be always allowed?

yuyichao commented:
Oh I guess `a.b @=(c.d=1)` is a syntax surgar for `a.b = a.b@(c.d=1)`?

iamed2 commented:
> The indexing chains have pretty complicated lowering, so they're quite a pain to write manually.

So changing this:
```julia
a.b = a.b@(c.d=1)
```
to this or something like it:
```julia
a.b = copy(a.b, :(c.d)=>2)
```
would be infeasible?

StefanKarpinski commented:
If we made the `copy` function magical, `a.b = copy(a.b, c.d = 2)` would be possible syntax. Although then it's not "just a function" anyway, but it does read more clearly to me.

iamed2 commented:
> If we made the copy function magical, a.b = copy(a.b, c.d = 2) would be possible syntax. Although then it's not "just a function" anyway, but it does read more clearly to me.

AFAIK there's already specific parser translations for keyword arguments turning `x = 2` into `(:x, 2)`. Maybe if `c.d` in that position parsed to `FieldAccess(:c, :d)` or something it would be doable without making `copy` magical? It wouldn't break existing code as anything not a symbol currently gives `syntax: keyword argument is not a symbol` and the code does actually parse right now.

```
julia> q = :(copy(a.b, c.d=2))
:(copy(a.b,c.d=2))

julia> dump(q)
Expr
  head: Symbol call
  args: Array{Any}((3,))
    1: Symbol copy
    2: Expr
      head: Symbol .
      args: Array{Any}((2,))
        1: Symbol a
        2: QuoteNode
          value: Symbol b
      typ: Any
    3: Expr
      head: Symbol kw
      args: Array{Any}((2,))
        1: Expr
          head: Symbol .
          args: Array{Any}((2,))
            1: Symbol c
            2: QuoteNode
              value: Symbol d
          typ: Any
        2: Int64 2
      typ: Any
  typ: Any

julia> eval(q)
------ ErrorException ------------------ Stacktrace (most recent call last)
 [1] — eval(::Any)
       ⌙ at sys.dylib:?
 [2] — eval(::Any)
       ⌙ at boot.jl:233
 [3] — eval(::Module, ::Any)
       ⌙ at sys.dylib:?
 [4] — eval(::Module, ::Any)
       ⌙ at boot.jl:234

syntax: keyword argument is not a symbol: "c.d"
```

Keno commented:
> The @ syntax without = will create a ref and assigning to it is semantically defined as a store to only that field's memory location.

setindex!(r, v), is yes.

> The @ syntax with = will create a whole new object (recursively mutating fields). The default implementatiion will be shared with the ref syntax using the default setfield.

Correct.

> x = x@([].a = 2) could be used instead of x@[].a = 2 when one of the type disallows gep explicitly

yes, but with the old caveat about data races, etc. The @(=) syntax is also useful for immutables bindings, which obviously can't be mutated as references.

> I hope it is still possible to emit x@[].a = 2 for a type that disallow gep for private implentation? (Possibly by calling the Core.gepfield directly)

I don't really have a problem with letting you put `let gepfield = Core.gepfield` and doing whatever you want here, pretty much like unsafe_store.

Keno commented:
I'm fine with doing something magical about copy, but @JeffBezanson was concerned about having too much verbosity/repetition in cases like `a.b = copy(a.b, c.d = 1)`.

andyferris commented:
What's `gep` in `gepfield` and `gepindex`? Get pointer?

Keno commented:
It's short for `getelementptr`, which is the LLVM operation that this maps to.

andyferris commented:
Thanks.

Not to impede progress here, but as a very gentle suggestion, could we first enable `RefField` as a feature along with `gepindex` and `gepfield` (though I would think something like `getindexref` and `getfieldref` may be clearer?), and later think about tailoring the surface-level syntax sugar towards the typical use cases that appear? The only reason is that I see 

> `a.b @=(c.d=1)`

and think, "Holy cow - that syntax!". (I mean, this *is* the same language that is simultaneously moving away from `&` as `Ref`, after all...).

Keno commented:
Well, I think it's fair to say that there's not that many place where you'd actually want that syntax (only for types that disallow unchecked field modification), so I don't mind not having the `@=` syntax. I do still kinda like the,
`update_green(a::RGB, ng) = a@(g = ng)`
syntax, because I feel like it jives well with the rest of the new proposed syntax.

StefanKarpinski commented:
FWIW, I'm not loving the `@` syntax either, but I've come around to other syntaxes before.

ararslan commented:
I absolutely love @iamed2's `copy` suggestion. I think it makes it crystal clear what's going on, whereas it took me quite a while to understand what was happening with the `@` stuff.

To me, that feels too "syntaxy," i.e. having a custom syntax for something kind of for the sake of having a custom syntax. Plus I imagine it creating confusion for new users, both because `a.b @=(c.d=1)` feels kinda "wat" and because it reuses the macro call operator.

StefanKarpinski commented:
A macro form `@copy! a.b c.d=v` could generate `a.b = copy(a.b, c.d = v)` or equivalent.

JeffBezanson commented:
I can understand the `@` being "too syntaxy". I think if we're going to use a function call, I'd rather just use `setindex` and `setfield` (no bang), since those are exactly what you'd implement for immutable data structures anyway. Most immutable data structures support `setindex`, but it might not copy anything.

Keno commented:
The lowering is to those function calls, but the problem is that the chains are annoying to generate, e.g. `a.b@(c.d = 1)` is `setfield(a.b, :c, setfield(a.b.c, :d, 1))`, etc. I do think we want a convenient way to write this.

Keno commented:
Could be a macro though I guess `@setfield(a.b, b.c = 1)`.

JeffBezanson commented:
The only prior art I'm aware of for this operation is something like `{ record | field = newvalue }` in functional languages. I don't know what they do for chaining fields though --- we should look into it and see if there are any ideas there.

johansigfrids commented:
This reminds me a lot of the spread operator from ES2017:
```JavaScript
const foo = {
    a: 1,
    b: 2,
    c: 3,
};
const newFoo = {
    ...foo,
    b: 9
};
console.log(foo) // Object { a: 1, b: 2, c: 3 }
console.log(newFoo) // Object {a: 1, b: 9, c: 3}
```
Crazy syntax proposal?: `a.b = (...a.b, c.d=1)`

StefanKarpinski commented:
The idea of splatting an object as a form of keyword argument into an object construction literal syntax is not that crazy, although to date we don't have any such syntax. I actually really like the idea of being able to use objects as storage for keyword arguments. People already do this fairly often, but they have to do the translation between keywords and objects manually. I can imagine writing something like this:

```jl
struct OrderOptions
    lt::Function
    by::Function
    rev::Bool
end

function sort!(v::AbstractVector; ord::OrderOptions...)
    # ...
    sort!(v, #= other args =#; ord...) # uses the fields of ord as key-values pairs
    # ...
end
```

StefanKarpinski commented:
This also relates to the original proposed syntax for constructing immutables with changed fields:

```jl
Foo(x, field=v)
```

See https://github.com/JuliaLang/julia/pull/6122, https://github.com/JuliaLang/julia/issues/5333.

Keno committed:
Unoptimized implementation of new semantics

Keno commented:
I've pushed an unoptimized implementation of the proposed new semantics (ref syntax + `@setfield` macro). Try it out and see if you like it. If so, I can do the optimization to make it fast.

StefanKarpinski commented on test/parse.jl:
```diff
@@ -1195,3 +1195,26 @@ end
+@test expand(parse("a@[c,d] = 1")) == :(setindex!(gepindex(a, c, d), 1))
+@test expand(parse("a@[c].d = 1")) == :(setindex!(gepfield(gepindex(a, c),:d), 1))
+
+expand(:(foo(1)@a))
```
???

Keno replied:
shouldn't crash, the LHS of an `@` can in principle be any expression. Writing the expression it expands to is a bit complicated though (has a bunch of SSAVals). I'll probably remove that here for the final version in favor of actually testing that it does what you expect semantically.

ararslan commented:
I'm still not at all a fan of using `@` as infix in this way. @yuyichao suggested that this could simply be a vararg macro. I imagine that this could be as simple as

```julia
struct A
    a::Int
    b::Float64
    c::String
end

macro setfield(x, args...)
    k = map(i->first(i.args), args)
    v = map(i->last(i.args), args)
    quote
        (typeof($x))([n in $k ? ($v)[findfirst($k, n)] : getfield($x, n) for n in fieldnames($x)]...)
    end
end

x = A(1, 2.0, "hello")
y = @setfield(x, b=3.0, c="hola")
```

This seems to work as expected. Surely the implementation can be optimized, but this seems like the simplest, most straightforward, easiest to understand approach, at least in terms of syntax. (I still don't quite understand what `a@b` is supposed to do, despite having read through this several times, but hopefully something as simple as the above would preclude the need for such a syntax.)

Keno commented:
`@setfield` is implemented as a macro now. The ref syntax needs to be shorter than that though, since it's gonna be quite commonly used.

ararslan commented:
> The ref syntax needs to be shorter than that though, since it's gonna be quite commonly used.

Can you elaborate more on how and why it will be used? I don't understand why `@setfield` couldn't just be the way to set the value and `object.field` could continue to be the way to access the fields.

Keno commented:
One of the two needs to have a nice infix syntax that can be used in assignment. Given that the preference that was expressed was for it to be the ref syntax because it doesn't have a data race. The use cases for both features outside of modifying an immutable field within some mutable value are somewhat marginal, so it's fine to relegate them to somewhat more obscure syntax.

Keno committed:
WIP

Keno committed:
Add extra early memcpyopt pass

Under certain circumstances, we emit loads/stores of large LLVM
structs. A lot of these can be trivially folded to memcpy and memcpyopt
is capable of doing so, but we weren't running it until after SROA.
SROA unfortunately, likes to take these apart, causing exponential
compile-time blow up and reduced runtime peroformance. In one particular
case (2000 element tuple), this change results in a 100x improvement
in compile time.

Keno committed:
Some tweaks based on experiments

Keno commented:
I've been asked to summarize the new semantics from the user perspective, so here goes. `x@f` takes a reference to field `f` in mutable object `x`. This references can be stored to/loaded from using getindex/setindex!, i.e.:
```
mutable struct foo
f::Tuple{Int}
end
x = foo((1,))
r = x@f[1] # Reference to f[1] in x
println(r[]) # Prints `1`
r[] = 2
println(x.f[1]) # Prints `2`
```
Additionally, if the `@` syntax occurs on the LHS of a (potentially updating) assignment, the `[]` is added implicitly, i.e.
```
x@f[1] += 1
```
is equivalent to
```
(x@f[1])[] = (x@f[1])[] + 1
```
Additionally, the latest changes makes RefValue behave transparently with respect to field references, i.e.
```
x = Ref{Tuple{Int}}()
r@[1] # Reference to the first element of the tuple - strictly this would have had to be r@[][1]
```
which also has the nice byproduct that `RefValue` and `RefField` behave equivalently with respect to their parameter type, e.g.
```
update_thing!(r::Ref{foo}, x) = r@f[1] = x
```
can take either a `Ref{foo}()` or a foo in some other mutable, by taking a reference to it.

Keno commented:
From the optimization point of view, the primary problem is that inference currently doesn't propagate the constants through inlining aggressively enough, i.e. for `gepfield(a, :x)`, it just treats the `:x` as a symbol, so it can't figure out what the field type is. @JeffBezanson @vtjnash thoughts on inference improvements to make that happen. I suppose we could use `gepfield(a, Val{:x}())` instead, but the plain symbol seems nicer.

stevengj commented:
See also #17115.  If I have an array `a` of an immutable type, does `a[i]@b = c` overwrite the entire `a[i]` element or just the `b` field?

yuyichao commented:
Only `b`. The two sides of the assignment determines the exact element that is being assigned to.

stevengj commented:
Now that we have constant propagation in inference, can this be revisited?

stevengj commented:
Bump?

smldis commented:
If possible, consider also mutating elements of vectors of immutables a la `vec[i]@x =1`

sairus7 commented:
There are several discussions on that, may be this helps:
https://discourse.julialang.org/t/make-immutable-mutable-again-or-make-arrays-of-structures-stack-allocated/20800

mauro3 commented:
Just to mention the package which implements this (with macros): https://github.com/jw3126/Setfield.jl with macros.

EricForgy commented:
> Just to mention the package which implements this (with macros): https://github.com/jw3126/Setfield.jl with macros.

Any thoughts on the lens approach? It seems pretty slick to me 😊

mohamed82008 commented:
Can this be part of 1.6?

vtjnash commented:
We should still work on this (#11902). It just needs some design work and agreement, and probably needs to be rewritten since this is stale.

AhmedSalih3d commented:
Out of curiosity, what happened to this proposal?

Did some package replace it or was it deemed not worth it any more?

mauro3 commented:
There is https://github.com/JuliaObjects/Accessors.jl
